plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java

```diff
@@ -39,6 +42,31 @@
     // Filter guaranteed to be enforced by Iceberg connector
     private final TupleDomain<IcebergColumnHandle> enforcedPredicate;
 
+    private final Optional<TableMetadata> icebergTableMetadata;
```

Comment:
Storing this in the table handle as a non-serializable property is not great. I know we do this in `HiveTableHandle`, but we should avoid it if possible. Instead, we should be able to have a map in `IcebergMetadata`. We need to think about how renames work, since `getTableHandle()` might be called twice in a query and resolve to different tables if a name was swapped. The table location should be good as an immutable identifier, so we might need to store that in the handle. Or perhaps we should simply cache calls to `getTableHandle()` so that we always resolve to the same handle within a transaction?

Reply:
Cached Iceberg's TableMetadata in IcebergMetadata.

Also removed the refactoring from first iteration that replaced `org.apache.iceberg.Table` with `org.apache.iceberg.TableMetadata`, so the diff is smaller now. Creating `org.apache.iceberg.Table` with an already present `org.apache.iceberg.TableMetadata` doesn't have much overhead, but may want to still cleanup in future to avoid creating/exposing the whole `Table` object even where not required.